### ISSUE TITLE
problem cause when doing w8a8kvfp8 for qwen2.5-vl

### ISSUE THREAD

**Issue by coolKeen** [vllm-project/llm-compressor#1354]
hi, I want to using w8a8kvfp8 for qwen2.5-vl and then run it on vllm. But when I doing quantization like the code below, I face the problem line the figure below, is there anything wrong with the recipe? Could you please give some example to get w8a8kvfp8 results of vision language model? Thanks in adavance !!! 

BTY, do you hace any document to explain the meaning of the parameters of QuantizationModifier?

![Image](https://github.com/user-attachments/assets/d383b084-ccaf-4bd2-b86a-0cb07a0fbf64)

`from transformers import AutoProcessor, Qwen2_5_VLForConditionalGeneration
from datasets import load_dataset

from llmcompressor import oneshot
from llmcompressor.modifiers.quantization import QuantizationModifier
import argparse
import os
import base64
from io import BytesIO
import torch
from datasets import load_dataset
from qwen_vl_utils import process_vision_info
from modelscope import AutoProcessor
from llmcompressor.modifiers.quantization import GPTQModifier
from llmcompressor.transformers import oneshot
from llmcompressor.transformers.tracing import (
    TraceableQwen2_5_VLForConditionalGeneration,
)
from compressed_tensors.quantization.quant_args import (
    QuantizationArgs,
    QuantizationStrategy,
    QuantizationType,
)

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--src_model_path", type=str)
    parser.add_argument("--dst_model_path", type=str) 
    args = parser.parse_args() 

    MODEL_ID = args.src_model_path
    dst_model_path = args.dst_model_path
    os.makedirs(dst_model_path, exist_ok=True)

    # Load model.
    model = Qwen2_5_VLForConditionalGeneration.from_pretrained(
        MODEL_ID, device_map="auto", torch_dtype="auto"
    )
    processor = AutoProcessor.from_pretrained(MODEL_ID)

    # Oneshot arguments
    DATASET_ID = "lmms-lab/flickr30k"
    DATASET_SPLIT = {"calibration": "test[:512]"}
    NUM_CALIBRATION_SAMPLES = 512
    MAX_SEQUENCE_LENGTH = 2048

    # Load dataset and preprocess.
    ds = ds = load_dataset(DATASET_ID, split=DATASET_SPLIT)

    ds = ds.shuffle(seed=42)

    dampening_frac=0.01

    def preprocess_and_tokenize(example):
        # preprocess
        buffered = BytesIO()
        example["image"].save(buffered, format="PNG")
        encoded_image = base64.b64encode(buffered.getvalue())
        encoded_image_text = encoded_image.decode("utf-8")
        base64_qwen = f"data:image;base64,{encoded_image_text}"
        messages = [
            {
                "role": "user",
                "content": [
                    {"type": "image", "image": base64_qwen},
                    {"type": "text", "text": "What does the image show?"},
                ],
            }
        ]
        text = processor.apply_chat_template(
            messages, tokenize=False, add_generation_prompt=True
        )
        image_inputs, video_inputs = process_vision_info(messages)

        # tokenize
        return processor(
            text=[text],
            images=image_inputs,
            videos=video_inputs,
            padding=False,
            max_length=MAX_SEQUENCE_LENGTH,
            truncation=True,
        )

    ds = ds.map(preprocess_and_tokenize, remove_columns=ds["calibration"].column_names)

    # Define a oneshot data collator for multimodal inputs.
    def data_collator(batch):
        assert len(batch) == 1
        return {key: torch.tensor(value) for key, value in batch[0].items()}

    recipe = [
        QuantizationModifier(scheme={
            "weights": QuantizationArgs(
                num_bits=8,
                type=QuantizationType.FLOAT,
                # group_size=128,
                strategy=QuantizationStrategy.TENSOR,
                symmetric=True,
                dynamic=False,
                ignore=["re:.*lm_head", "re:visual.*"],
                targets=["Linear"]

            ),
            "input_activations": QuantizationArgs(
                num_bits=8,
                type=QuantizationType.FLOAT,
                strategy=QuantizationStrategy.TENSOR,
                symmetric=True,
                dynamic=True,
                observer=None,
                ignore=["re:.*lm_head", "re:visual.*"],
                targets=["Linear"]
            ),
            "kv_cache_scheme": QuantizationArgs(
                num_bits=8,
                type=QuantizationType.FLOAT,
                strategy=QuantizationStrategy.TENSOR,
                symmetric=True,
                dynamic=True,
                observer=None,
                ignore=["re:.*lm_head", "re:visual.*"],
                targets=["Linear"]
            )
        })
    ]

    oneshot(
        model=model,
        dataset=ds,
        recipe=recipe,
        max_seq_length=MAX_SEQUENCE_LENGTH,
        num_calibration_samples=NUM_CALIBRATION_SAMPLES,
    )

    SAVE_DIR = dst_model_path
    oneshot(model=model, recipe=recipe, output_dir=SAVE_DIR)
    processor.save_pretrained(SAVE_DIR)
`

**Comment by kylesayrs** [comment]
Hi @coolKeen! Thanks for using LLM Compressor!

The format used by Compressed Tensors is slightly different than what you've specified, namely that the kv_cache_scheme is a field of `QuantizationModifier`, not of `QuantizationScheme`.

```python3
recipe = [
    QuantizationModifier(
        scheme={
            "weights": QuantizationArgs(
                num_bits=8,
                type=QuantizationType.FLOAT,
                # group_size=128,
                strategy=QuantizationStrategy.TENSOR,
                symmetric=True,
                dynamic=False,
                ignore=["re:.*lm_head", "re:visual.*"],
                targets=["Linear"]
            ),
            "input_activations": QuantizationArgs(
                num_bits=8,
                type=QuantizationType.FLOAT,
                strategy=QuantizationStrategy.TENSOR,
                symmetric=True,
                dynamic=True,
                observer=None,
                ignore=["re:.*lm_head", "re:visual.*"],
                targets=["Linear"]
            ),
        },
        kv_cache_scheme=QuantizationArgs(
            num_bits=8,
            type=QuantizationType.FLOAT,
            strategy=QuantizationStrategy.TENSOR,
            symmetric=True,
            dynamic=True,
            observer=None,
            ignore=["re:.*lm_head", "re:visual.*"],
            targets=["Linear"]
        )
    )
]
```

Should should be a valid recipe, however there are a few bugs with handling this.

The first is that there is a bug with the [config parsing logic](https://github.com/vllm-project/llm-compressor/blob/main/src/llmcompressor/modifiers/quantization/quantization/base.py#L146) which is leading to the error you're seeing.

The second is that kv_cache_schemes with dynamic quantization are not being parsed correctly. If you want dynamic quantization, simply remove the kv_cache_scheme for now

In the meantime, this recipe should work for your use case

```python3
recipe = [
    QuantizationModifier(
        config_groups={
            "group_0": dict(
                targets=["Linear"],
                weights=QuantizationArgs(
                    num_bits=8,
                    type=QuantizationType.FLOAT,
                    # group_size=128,
                    strategy=QuantizationStrategy.TENSOR,
                    symmetric=True,
                    dynamic=False,
                    ignore=["re:.*lm_head", "re:visual.*"],
                    targets=["Linear"]
                ),
                input_activations=QuantizationArgs(
                    num_bits=8,
                    type=QuantizationType.FLOAT,
                    strategy=QuantizationStrategy.TENSOR,
                    symmetric=True,
                    dynamic=True,
                    observer=None,
                    ignore=["re:.*lm_head", "re:visual.*"],
                    targets=["Linear"]
                )
            )
        },
    )
]
```

**Comment by coolKeen** [comment]
> Hi [@coolKeen](https://github.com/coolKeen)! Thanks for using LLM Compressor!
> 
> The format used by Compressed Tensors is slightly different than what you've specified, namely that the `scheme` field maps preset names to targets, and that kv_cache_scheme is a field of `QuantizationModifier`, not of `QuantizationScheme`.
> 
> Beyond this, there is a validation error where kv_cache_schemes with dynamic quantization are not being parsed correctly. If you want dynamic quantization, simply remove the kv_cache_scheme for now.
> 
> In the meantime, this recipe should work for your use case
> 
> recipe = [
>     QuantizationModifier(
>         config_groups={
>             "group_0": dict(
>                 targets=["Linear"],
>                 weights=QuantizationArgs(
>                     num_bits=8,
>                     type=QuantizationType.FLOAT,
>                     # group_size=128,
>                     strategy=QuantizationStrategy.TENSOR,
>                     symmetric=True,
>                     dynamic=False,
>                     ignore=["re:.*lm_head", "re:visual.*"],
>                     targets=["Linear"]
>                 ),
>                 input_activations=QuantizationArgs(
>                     num_bits=8,
>                     type=QuantizationType.FLOAT,
>                     strategy=QuantizationStrategy.TENSOR,
>                     symmetric=True,
>                     dynamic=True,
>                     observer=None,
>                     ignore=["re:.*lm_head", "re:visual.*"],
>                     targets=["Linear"]
>                 )
>             )
>         },
>     )
> ]

@kylesayrs thanks for you respond!! But I also want to use fp8 for kv cache and run it on vllm. How should I change this recipe.

**Comment by coolKeen** [comment]
@kylesayrs I change recipe to 
    recipe = [
    QuantizationModifier(
        config_groups={
            "group_0": dict(
                targets=["Linear"],
                weights=QuantizationArgs(
                    num_bits=8,
                    type=QuantizationType.FLOAT,
                    # group_size=128,
                    strategy=QuantizationStrategy.TENSOR,
                    symmetric=True,
                    dynamic=False,
                    ignore=["re:.*lm_head", "re:visual.*"],
                    targets=["Linear"]
                ),
                input_activations=QuantizationArgs(
                    num_bits=8,
                    type=QuantizationType.FLOAT,
                    strategy=QuantizationStrategy.TENSOR,
                    symmetric=True,
                    dynamic=True,
                    observer=None,
                    ignore=["re:.*lm_head", "re:visual.*"],
                    targets=["Linear"]
                )
            )
        },
        kv_cache_scheme=QuantizationArgs(
                    num_bits=8,
                    type=QuantizationType.FLOAT,
                    strategy=QuantizationStrategy.TENSOR,
                    symmetric=True,
                    dynamic=False,
                    observer=None,
                    ignore=["re:.*lm_head", "re:visual.*"],
                    targets=["Linear"]
                )
    )
]

but another problem cause:

![Image](https://github.com/user-attachments/assets/ee7f9732-5430-480d-a233-356ef932f936)
looks like something wrong with kv cache fp8 calibration?

**Comment by coolKeen** [comment]
btw，what is the difference between static and dynamic for fp8?

**Comment by liuy-2019** [comment]
Dynamic fp8 : When the weight is loaded, it will dynamically calculate the scale, then quantify it to fp8, and then load it into memory? Is this right? @kylesayrs

**Comment by liuy-2019** [comment]
@coolKeen  I meet the same quastion, because batch["image_grid_thw"] = torch.tensor([[1, 16, 16]]) should be edited as batch["image_grid_thw"] = torch.tensor([1, 16, 16])

**Comment by liuy-2019** [comment]
> [@coolKeen](https://github.com/coolKeen) I meet the same quastion, because batch["image_grid_thw"] = torch.tensor([[1, 16, 16]]) should be edited as batch["image_grid_thw"] = torch.tensor([1, 16, 16])

There is a new solution:

```
def data_collator(batch):
    assert len(batch) == 1
    return {key: torch.tensor(value) for key, value in batch[0].items()}

oneshot(
    model=model,
    tokenizer=LOCAL_MODEL_PATH,
    dataset=ds,
    recipe=recipe,
    max_seq_length=MAX_SEQUENCE_LENGTH,
    num_calibration_samples=NUM_CALIBRATION_SAMPLES,
    data_collator=data_collator,
    output_dir=dst_model_path
)
```

**Comment by kylesayrs** [comment]
@coolKeen @liuy-2019 

> Dynamic fp8 : When the weight is loaded, it will dynamically calculate the scale, then quantify it to fp8, and then load it into memory? Is this right? @kylesayrs 

Yes, that's correct. Dynamic means that scales (quantization parameters) are computed at runtime, which comes at a performance cost but higher accuracy and no calibration data requirement.

**Comment by kylesayrs** [comment]
@liuy-2019 

Your data collator solution is correct!
